File: DSA/3Recursion/Recursion3/ropeCuttingProblem.py
# Given a rope length L and 3 numbers a, b, c
# cut the rope into max no of pieces s.t. the length of each piece is a/b/c
# return the max no of pieces

def get_no_of_pieces_helper(L, a, b, c, s, pieces):
    if s > L:
        return 0
    
    if s == L:
        return pieces
    
    x = get_no_of_pieces_helper(L, a, b, c, s+a, pieces+1)
    y = get_no_of_pieces_helper(L, a, b, c, s+b, pieces+1)
    z = get_no_of_pieces_helper(L, a, b, c, s+c, pieces+1)
    
    return max(x, y, z)

def get_no_of_pieces(L, a, b, c):
     return get_no_of_pieces_helper(L, a, b, c, 0, 0)


L = abs(int(input("Enter the length of rope: ")))
a = abs(int(input("Enter the length of piece a: ")))
b = abs(int(input("Enter the length of piece b: ")))
c = abs(int(input("Enter the length of piece c: ")))

print("Max no of pieces:", get_no_of_pieces(L, a, b, c))

File: DSA/3Recursion/Recursion3/RopeCuttingProblem.java
import java.util.Scanner;

public class RopeCuttingProblem {
    public static int getNoOfPiecesHelper(int L, int a, int b, int c, int extracted, int pieces) {
        if (extracted > L) {
            return 0;
        }

        if (extracted == L) {
            return pieces;
        }

        int p1 = getNoOfPiecesHelper(L, a, b, c, extracted+a, pieces+1);
        int p2 = getNoOfPiecesHelper(L, a, b, c, extracted+b, pieces+1);
        int p3 = getNoOfPiecesHelper(L, a, b, c, extracted+c, pieces+1);

        return Math.max(Math.max(p1, p2), p3);
    }

    public static int getNoOfPieces(int L, int a, int b, int c) {
        return getNoOfPiecesHelper(L, a, b, c, 0, 0);
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        System.out.print("Enter length of rope: ");
        int L = Math.abs(scan.nextInt());

        System.out.print("Enter length of piece a, b, c (seperated by space or nextLine): ");
        int a = Math.abs(scan.nextInt());
        int b = Math.abs(scan.nextInt());
        int c = Math.abs(scan.nextInt());
        scan.close();

        System.out.println("The max no of pieces = " + getNoOfPieces(L, a, b, c));
    }
}
